Federation: sanitize paths to avoid path traversal attacks

jschaul opened the pull request:
At the moment we use the convention that all RPCs between federator -> brig/galley/etc use the "federation" path prefix. This PR adds validation to ensure other paths cannot be called, as that would expose non-federation APIs (including internal, non-authenticated ones) to the world.

See also the discussion in https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/224166764/Limiting+access+to+federation+endpoints

## Checklist

 - [x] Title of this PR explains the impact of the change.
 - [x] The description provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] rebase to develop after #1637 has landed.
 - [x] The CHANGELOG.md file in the *Unreleased* section has been updated to explain the change which will be included in the release notes.

jschaul commented on services/federator/test/integration/Test/Federator/InwardSpec.hs:
```diff
@@ -84,6 +84,24 @@ spec env =
+      runTestFederator env $ do
+        let o = object ["name" .= ("fakeNewUser" :: Text)]
+        err <- asInwardErrorUnsafe <$> inwardBrigCall "federation/../i/users" (encode o)
+        expectErr IForbiddenEndpoint err
```
dear reviewers, I'd love ideas on what other things to test. Would any property tests make sense here?

pcapriotti replied:
I think having property tests for `sanitizePath` would be useful, but of course only insofar as the generator produces realistic URL inputs. As for ad-hoc test cases, you could maybe add some with non-ascii character (both percent-encoded and not), query string and/or fragments.

mdimjasevic replied:
I agree, some (unit) tests for `sanitizePath` would be good to have, including one that has "federation" as the only path segment.

jschaul replied:
I added a bunch of unit tests, take a look.

pcapriotti replied:
Nice! :+1:

kkohbrok reviewed (commented):
Looks solid already! Another common vector for attacks would be encoding. I assume that `URIB.parseRelativeRef` handles the decoding of the byte string, so we could test the normalization function for its handling of [percent encoding](https://en.wikipedia.org/wiki/Directory_traversal_attack#Percent_encoding_in_URIs). I don't have a lot of experience in webapp security, but it seems to be a common enough attack vector to have some tests for it, i.e. to make sure that the normalizer properly handles percent decoding before sanitization.

pcapriotti commented on services/federator/src/Federator/Validation.hs:
```diff
@@ -60,3 +63,30 @@ validateDomain unparsedDomain = do
+  let normalized = URIB.normalizeURIRef' URIB.aggressiveNormalization uriRef
+  -- next, we use the Network.URI library to get path segments after normalization,
+  -- and ensure the first segment in the path is "federation"
+  uri <- case parseURIReference (cs normalized) of
```
Here you are converting the parsed URI back to a `String`, then reparsing it. Wouldn't it be better to just take the path (as `ByteString`) and directly check that it begins with (ascii-encoded) "federation/"?

jschaul replied:
Done.

pcapriotti reviewed (approved):
Looks good!

mdimjasevic commented on services/federator/test/integration/Test/Federator/InwardSpec.hs:
```diff
@@ -84,6 +84,24 @@ spec env =
         err <- asInwardErrorUnsafe <$> inwardBrigCall "federation/this-endpoint-does-not-exist" (encode Aeson.emptyObject)
         expectErr IInvalidEndpoint err
 
+    it "should only not accept invalid paths" $
```
```suggestion
    it "should not accept invalid paths" $
```

mdimjasevic reviewed (approved):
It looks good. I haven't checked, but I suppose `sanitizePath` is already used for all federator paths?

jschaul commented:
I added a range of tests, and also changed the implementation to guard against more cases. Merging this PR now, but in case someone still has other ideas, feel free to comment.
